### ISSUE TITLE
tf.control_dependencies() not respected with Keras layers

### ISSUE THREAD

**Issue by humcasma** [tensorflow/tensorflow#24384]
**System information**
- TensorFlow version: v1.12.0-0-ga6d8ffae09 1.12.0
- Python version: 3.6.7

I want to switch dropout on and off while processing one single batch of data (i.e., feeding data just once). For this I replace the _learning_phase_ placeholder used by Keras with a variable, which I can then assign dynamically, and use `tf.control_dependencies()` to control the order of operations. Order of execution does not seem to be respected, leading to random behavior.
More details: In the context of reinforcement learning, for each batch of data, I want to: (1) do a forward pass over my network with dropout disable to build targets; and (2) do a training step with those targets with dropout enabled.   

**Code to reproduce the issue**
The code below is a toy example that resembles what I want to do. I would expect `o1` and `o3` to be equal and different than `o2`. I would also expect to run w/o crashing. 

```
from tensorflow.keras.layers import Dropout, Dense, Input, Lambda, Layer, Add, RepeatVector, LSTM
from tensorflow.python.keras import backend as K
from tensorflow.keras import Model
import tensorflow as tf
import numpy as np

lph = tf.Variable(False, dtype=tf.bool, name='lph')
g = tf.get_default_graph()
K._GRAPH_LEARNING_PHASES[g] = lph

repeat = RepeatVector(2)
lstm = LSTM(1, dropout=0.5, recurrent_dropout=0.5)

ph1 = tf.placeholder(tf.float32, shape=(None, 1), name='ph1')
h = repeat(ph1)
o1 = lstm(h)

with g.control_dependencies([lph.assign(True), tf.identity(o1)]):
  target1 = tf.stop_gradient(o1)

  ph2 = tf.placeholder(tf.float32, shape=(None, 1), name='ph2')
  h = repeat(ph2)
  o2 = lstm(h)
  
  with g.control_dependencies([lph.assign(False), tf.identity(o2)]):
    target2 = tf.stop_gradient(o2)

    ph3 = tf.placeholder(tf.float32, shape=(None, 1), name='ph3')
    h = repeat(ph3)
    o3 = lstm(h)
    
    loss = target1 + target2 - o2
    opt = tf.train.GradientDescentOptimizer(0.1)
    train = opt.minimize(loss)

sess = tf.Session()
K.set_session(sess)
sess.run(tf.global_variables_initializer())

data = np.ones((1,1))
sess.run([o1, o2, o3, train], feed_dict={ph1:data, ph2:data, ph3:data})
```

**Comment by humcasma** [comment]
Closing this since there is not a problem, actually. Splitting each of the `control_dependencies()` commands in two, as shown below, solves the issue.

```
with g.control_dependencies([o1]):
       with g.control_dependencies([lph.assign(True)]):
               ....
```
